### ISSUE TITLE
autodoc: inconsistent cross reference targets for class attributes

### ISSUE THREAD

**Issue by mwiebusch78** [sphinx-doc/sphinx#1672]
There seems to be an inconsistency with cross reference targets for class attributes generated by autodoc. When I refer to method f of class C in module mod the cross reference target is mod.C.f, but for an attribute attr in the same class it is C.attr (i.e. without the leading mod). This is inconsistent and could lead to clashes when you have several modules in the game.

**Comment by birkenfeld** [comment]
I'll have a look.

**Comment by birkenfeld** [comment]
Hmm, I can't reproduce this.

**Comment by birkenfeld** [comment]
Since I couldn't reproduce this, I'll close it now.

**Comment by cdunklau** [comment]
EDIT: This doesn't appear to be the same issue, my problem appears to stem from napoleon, not autodoc. Investigation below remains.

I ran into this. https://github.com/cdunklau/sphinxattrissue has a not-so-minimal reproduction. @birkenfeld could you reopen and investigate? I spent a few hours on this, and I'm at a loss for why it's happening.

This commit cdunklau/sphinxattrissue@eff18e023250459d331a42c56da9d8b09a9bb908 shows the issue.

I fixed it by adding `.. module:: mypkg.models` before the autoclass directive, as shown in commit cdunklau/sphinxattrissue@773d7866eb67350d7c9d93c534f3b6bec74daf8e, but I'd really like to know why this happens. I'd expect `:attr:` refs to work the same as `:class:` and `:meth:` refs.

**Comment by cdunklau** [comment]
I've narrowed this down. I no longer believe this has anything to do with autodoc or napoleon.

cdunklau/sphinxattrissue@e6a9bca0a1c218c5bbfdb8d35025da6de0815da1 has a minimal reproduction, details are in minimal/README.rst

The markup itself:

```rst
.. attrissue documentation master file, created by
   sphinx-quickstart on Mon May 22 12:55:03 2017.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

Welcome to attrissue's documentation!
=====================================

Full documentation of the attributes is in the API documentation for
:class:`~mypkg.models.SomeClass` objects.

.. note::

    The links for the below attributes don't get created

The attributes that are likely the most useful are:

-   :attr:`~mypkg.models.SomeClass.id`: ID
-   :attr:`~mypkg.models.SomeClass.workflow_state`: Check the workflow state
-   :attr:`~mypkg.models.SomeClass.age`: How long ago
    (in milliseconds) the thing was last seen online.
-   :attr:`~mypkg.models.SomeClass.name`: The name of the thing
-   :attr:`~mypkg.models.SomeClass.host`: The IP address or hostname


.. note::

    But class and method refs do work properly...

:class:`~mypkg.models.SomeClass` instances also have a
:meth:`~mypkg.models.SomeClass.some_method` method.


API Reference
=============

SomeClass
+++++++++

..  without this module directive, the above links for the attributes don't get
    created, but the links for the class and the method do!
    
    .. module:: mypkg.models

.. This is the actual markup generated by autodoc and napoleon in the original
   reproduction attempt

.. py:class:: SomeClass(config)
   :module: mypkg.models

   Some class that does a thing
   
   .. attribute:: id
   
      *str* -- The identifier
   
   .. attribute:: workflow_state
   
      *int* -- The workflow state
   
   .. attribute:: age
   
      *int* -- Age in milliseconds
   
   .. attribute:: name
   
      *str* -- The name of the thing
   
   .. attribute:: host
   
      *str* -- The hostname or IP address
   
   
   .. py:method:: SomeClass.some_method()
      :module: mypkg.models
   
      This shows that method refs work



Table of contents
=================

.. toctree::
   :maxdepth: 2
   :caption: Contents:

**Comment by cdunklau** [comment]
The root cause of the issue appears to be that for `.. py:method::`, the `options` attribute of the directive has the module included (`{'module': 'mypkg.models'}`), but for `..py:attribute::`, the `options` attribute is an empty dict. I'm continuing to investigate.

**Comment by cdunklau** [comment]
And that's because `:module:` is there for the method, but not the attributes. My issue appears to be because of napoleon, not autodoc. With autodoc, this code:

```python
class AutodoccedClass(object):
    """
    A class to be documented with autodoc

    """
    #: Some documented attribute
    someattr = 123
    def method(self):
        """
        A method on the autodocced class
        """
```

Results in this generated output:

```rst
.. py:class:: AutodoccedClass
   :module: mymodule

   A class to be documented with autodoc
   
   
   
   .. py:method:: AutodoccedClass.method()
      :module: mymodule
   
      A method on the autodocced class
      
   
   .. py:attribute:: AutodoccedClass.someattr
      :module: mymodule
      :annotation: = 123
   
      Some documented attribute
```

So that's all working. It looks like napoleon isn't adding the `:module:` bit to the attributes.

**Comment by RobRuana** [comment]
Ah, interesting. I will look into this!

**Comment by cdunklau** [comment]
@RobRuana in my not-so-minimal reproduction here https://github.com/cdunklau/sphinxattrissue/tree/master/docs you can see the conf settings I have set for napoleon. I have a feeling that `napoleon_use_ivar = False` is required to reproduce this.

**Comment by RobRuana** [comment]
Thanks for the extra info. I have been particularly busy at work, so I may not be able to get to this issue until this weekend or the next.

**Comment by cdunklau** [comment]
no problem. I just tested that assumption, and no, it looks like it happens regardless of how `napoleon_use_ivar` is set.

**Comment by justinfx** [comment]
Any update on this issue? I am currently fighting to resolve build warnings because the use of `Attributes:` in a class docstring is causing a bunch of "reference target not found" issues.

**Comment by RobRuana** [comment]
I have absolutely no free time at this point. If someone else could look into this/submit a pull request, I would be very grateful.

**Comment by tk0miya** [comment]
It seems the latest Sphinx does not emit warnings at all on building https://github.com/cdunklau/sphinxattrissue. It seems something was fixed at v3.0.0 (I confirmed the warnings with v2.4.4).

So I'm closing this now. Please let me know if you still have the same (or similar) issue.
Thanks,
